test_fms/coupler/test_coupler_types.F90

```diff
@@ -0,0 +1,311 @@
+                                                                 !! initalization
+integer :: id_x, id_y, id_z, chksum_unit
+character(len=128) :: chksum_2d, chksum_3d
+real(FMS_CP_TEST_KIND_), allocatable :: expected_2d(:,:), expected_3d(:,:,:)
+integer :: err, ncid, dim1D, varid, day
+
+call fms_init
+call time_manager_init
+call fms2_io_init
+call mpp_init
+call set_calendar_type(JULIAN)
+
+! basic domain set up
+nlat=60; nlon=60; nz=12
+layout = (/2, 2/)
+call mpp_domains_set_stack_size(86400)
+call mpp_define_domains( (/1,nlon,1,nlat/), layout, Domain, name='test_coupler')
+call mpp_define_io_domain(Domain, (/1,1/))
+call mpp_get_data_domain(Domain, data_grid(1), data_grid(2), data_grid(3), data_grid(4))
+
+! create/allocate new types with routines in utils file
+call set_up_1d_coupler_type(bc_1d_new, data_grid)
+call set_up_2d_coupler_type(bc_2d_new, data_grid, appendix="new", to_read=.false.)
+data_grid(5) = nz
+call set_up_3d_coupler_type(bc_3d_new, data_grid, appendix="new", to_read=.false.)
+
+! coupler_type_set_data
+allocate(array_2d(data_grid(1):data_grid(2), data_grid(3):data_grid(4)))
+allocate(array_3d(data_grid(1):data_grid(2), data_grid(3):data_grid(4), data_grid(5)))
+array_2d = 1.0_lkind
+array_3d = 1.0_lkind
+do i=1, num_bc
+  do j=1, num_fields
+    call coupler_type_set_data(array_2d, i, j, bc_2d_new)
+    call coupler_type_set_data(array_2d, i, j, data_grid(5), bc_3d_new)
+    call coupler_type_set_data(array_3d, i, j, bc_3d_new)
+  enddo
+enddo
+call check_field_data_2d(bc_2d_new, array_2d)
+call check_field_data_3d(bc_3d_new, array_3d)
+
+! coupler_type_write_chksum
+! needs to write to a unit num
+call coupler_type_write_chksums(bc_2d_new, stdout())
+call coupler_type_write_chksums(bc_3d_new, stdout())
+
+! coupler_type_increment_data
+! creates copies to increment into original
+call coupler_type_copy(bc_2d_new, bc_2d_cp, data_grid(1), data_grid(2), data_grid(3), data_grid(4), " ", &
+                       (/ 0 /), time_t )
+call coupler_type_copy_data(bc_2d_new, bc_2d_cp)
+call coupler_type_copy(bc_3d_new, bc_3d_cp, data_grid(1), data_grid(2), data_grid(3), data_grid(4), data_grid(5), " ", &
+                       (/ 0 /), time_t )
+call coupler_type_copy_data(bc_3d_new, bc_3d_cp)
+call coupler_type_increment_data(bc_2d_new, bc_2d_cp)
+call coupler_type_increment_data(bc_3d_new, bc_3d_cp)
+! copy of itself incremented should just be 2.0
+array_2d = 2.0_lkind; array_3d = 2.0_lkind
+call check_field_data_2d(bc_2d_cp, array_2d)
+call check_field_data_3d(bc_3d_cp, array_3d)
+
+! coupler_type_rescale_data
+call coupler_type_rescale_data(bc_2d_cp, 2.0_lkind)
+call coupler_type_rescale_data(bc_3d_cp, 2.0_lkind)
+array_2d = 4.0_lkind; array_3d = 4.0_lkind ! data was 2, rescaled by factor of 2
+call check_field_data_2d(bc_2d_cp, array_2d)
+call check_field_data_3d(bc_3d_cp, array_3d)
+call coupler_type_destructor(bc_2d_cp)
```

Comment:
I feel like testing the destructor might be important because of bc and bc_r4.  Is there a way to check that the arrays were deallocated?